File: ios/WindBrassTuner/WindBrassTuner/PitchDetector.swift
@preconcurrency import AVFoundation
import Foundation

enum MicrophonePermissionState {
    case unknown
    case granted
    case denied
}

final class PitchDetector: ObservableObject, @unchecked Sendable {
    @Published private(set) var detectedFrequency: Double?
    @Published private(set) var inputLevel: Double = 0
    @Published private(set) var isRunning = false
    @Published private(set) var microphonePermission: MicrophonePermissionState = .unknown
    @Published private(set) var statusMessage = "Ready"

    private let engine = AVAudioEngine()
    private var smoothedFrequency: Double?
    private var missedFrameCount = 0

    func start(targetFrequency: Double) {
        let session = AVAudioSession.sharedInstance()

        switch session.recordPermission {
        case .granted:
            microphonePermission = .granted
            startEngine(targetFrequency: targetFrequency)
        case .denied:
            microphonePermission = .denied
            detectedFrequency = nil
            inputLevel = 0
            isRunning = false
            statusMessage = "Microphone access is off"
        case .undetermined:
            microphonePermission = .unknown
            statusMessage = "Requesting microphone access"
            session.requestRecordPermission { [weak self] granted in
                DispatchQueue.main.async {
                    guard let self else { return }

                    if granted {
                        self.microphonePermission = .granted
                        self.startEngine(targetFrequency: targetFrequency)
                    } else {
                        self.microphonePermission = .denied
                        self.detectedFrequency = nil
                        self.inputLevel = 0
                        self.isRunning = false
                        self.statusMessage = "Microphone access is off"
                    }
                }
            }
        @unknown default:
            microphonePermission = .denied
            detectedFrequency = nil
            inputLevel = 0
            isRunning = false
            statusMessage = "Microphone permission is unavailable"
        }
    }

    func stop() {
        engine.inputNode.removeTap(onBus: 0)
        engine.stop()
        engine.reset()

        resetStabilityState()
        detectedFrequency = nil
        inputLevel = 0
        isRunning = false
        statusMessage = "Ready"
    }

    private func startEngine(targetFrequency: Double) {
        if isRunning {
            stop()
        }

        resetStabilityState()
        configureAudioSession()

        let inputNode = engine.inputNode
        let format = inputNode.outputFormat(forBus: 0)

        inputNode.removeTap(onBus: 0)
        inputNode.installTap(onBus: 0, bufferSize: 4096, format: format) { [weak self] buffer, _ in
            let result = Self.process(
                buffer: buffer,
                sampleRate: buffer.format.sampleRate,
                targetFrequency: targetFrequency
            )

            DispatchQueue.main.async { [weak self] in
                self?.inputLevel = result.level
                self?.publishStableFrequency(result.frequency)
            }
        }

        do {
            try engine.start()
            isRunning = true
            statusMessage = "Listening..."
        } catch {
            detectedFrequency = nil
            isRunning = false
            statusMessage = "Could not start microphone"
        }
    }

    private func publishStableFrequency(_ frequency: Double?) {
        guard let frequency, frequency > 0 else {
            missedFrameCount += 1

            if missedFrameCount <= 8, let smoothedFrequency {
                detectedFrequency = smoothedFrequency
                statusMessage = "Holding tone"
            } else {
                detectedFrequency = nil
                statusMessage = missedFrameCount > 24 ? "No input detected" : "Listening..."
            }

            return
        }

        missedFrameCount = 0

        if let smoothedFrequency {
            let distanceInCents = abs(1200 * log2(frequency / smoothedFrequency))

            if distanceInCents < 700 {
                self.smoothedFrequency = (smoothedFrequency * 0.72) + (frequency * 0.28)
            } else {
                self.smoothedFrequency = frequency
            }
        } else {
            smoothedFrequency = frequency
        }

        detectedFrequency = smoothedFrequency
        statusMessage = "Pitch detected"
    }

    private func resetStabilityState() {
        smoothedFrequency = nil
        missedFrameCount = 0
    }

    private func configureAudioSession() {
        let session = AVAudioSession.sharedInstance()

        do {
            try session.setCategory(.playAndRecord, mode: .measurement, options: [.defaultToSpeaker])
            try? session.setPreferredSampleRate(44_100)
            try? session.setPreferredIOBufferDuration(0.02)
            if let builtInMic = session.availableInputs?.first(where: { $0.portType == .builtInMic }) {
                try? session.setPreferredInput(builtInMic)
            }
            try session.setActive(true)
        } catch {
            // The UI remains usable even if the audio session cannot be configured.
            statusMessage = "Audio session setup failed"
        }
    }

    nonisolated private static func process(
        buffer: AVAudioPCMBuffer,
        sampleRate: Double,
        targetFrequency: Double
    ) -> (frequency: Double?, level: Double) {
        guard let channel = buffer.floatChannelData?[0] else {
            return (nil, 0)
        }

        let frameLength = Int(buffer.frameLength)
        guard frameLength > 0 else {
            return (nil, 0)
        }

        let samples = Array(UnsafeBufferPointer(start: channel, count: frameLength))
        let rms = sqrt(samples.reduce(0) { $0 + Double($1 * $1) } / Double(samples.count))
        let level = min(1, rms * 22)

        guard rms > 0.0025 else {
            return (nil, level)
        }

        return (
            Self.detectPitch(samples: samples, sampleRate: sampleRate, targetFrequency: targetFrequency),
            level
        )
    }

    nonisolated static func detectPitch(samples: [Float], sampleRate: Double, targetFrequency: Double) -> Double? {
        guard samples.count > 32, sampleRate > 0, targetFrequency > 0 else {
            return nil
        }

        let minimumFrequency = 70.0
        let maximumFrequency = 1_200.0
        let minLag = max(8, Int(sampleRate / maximumFrequency))
        let maxLag = min(samples.count - 2, Int(sampleRate / minimumFrequency))

        guard minLag < maxLag else {
            return nil
        }

        let mean = samples.reduce(0.0) { $0 + Double($1) } / Double(samples.count)
        let centeredSamples = samples.map { Double($0) - mean }
        var scores = Array(repeating: 0.0, count: maxLag + 1)
        var bestLag = minLag
        var bestScore = -Double.infinity

        for lag in minLag...maxLag {
            var correlation = 0.0
            var energyA = 0.0
            var energyB = 0.0
            let limit = samples.count - lag

            for index in 0..<limit {
                let a = centeredSamples[index]
                let b = centeredSamples[index + lag]
                correlation += a * b
                energyA += a * a
                energyB += b * b
            }

            let denominator = sqrt(energyA * energyB)
            guard denominator > 0 else {
                continue
            }

            let score = correlation / denominator
            scores[lag] = score

            if score > bestScore {
                bestScore = score
                bestLag = lag
            }
        }

        guard bestScore > 0.45 else {
            return nil
        }

        var refinedLag = Double(bestLag)

        if bestLag > minLag, bestLag < maxLag {
            let left = scores[bestLag - 1]
            let center = scores[bestLag]
            let right = scores[bestLag + 1]
            let curvature = left - (2 * center) + right

            if abs(curvature) > 0.000_001 {
                let adjustment = 0.5 * (left - right) / curvature
                refinedLag += max(-0.5, min(0.5, adjustment))
            }
        }

        return sampleRate / refinedLag
    }
}

File: android/WindBrassTuner/app/src/main/java/com/olive/windbrasstuner/audio/PitchDetector.kt
package com.olive.windbrasstuner.audio

import android.annotation.SuppressLint
import android.media.AudioFormat
import android.media.AudioRecord
import android.media.MediaRecorder
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlin.concurrent.thread
import kotlin.math.abs
import kotlin.math.log2
import kotlin.math.max
import kotlin.math.min
import kotlin.math.sqrt
import java.util.concurrent.atomic.AtomicInteger

data class PitchDetectorState(
    val detectedFrequency: Double? = null,
    val isRunning: Boolean = false,
    val statusMessage: String = "Ready",
)

class PitchDetector {
    private val sampleRate = 44_100
    private val frameSize = 4096
    private val _state = MutableStateFlow(PitchDetectorState())
    val state: StateFlow<PitchDetectorState> = _state.asStateFlow()
    private val generation = AtomicInteger(0)

    @Volatile private var running = false
    private var audioRecord: AudioRecord? = null
    private var smoothedFrequency: Double? = null
    private var missedFrameCount = 0

    @SuppressLint("MissingPermission")
    fun start(targetFrequency: Double) {
        stop()
        val session = generation.incrementAndGet()
        val minimumBuffer = AudioRecord.getMinBufferSize(
            sampleRate,
            AudioFormat.CHANNEL_IN_MONO,
            AudioFormat.ENCODING_PCM_FLOAT,
        )
        if (minimumBuffer <= 0) {
            _state.value = PitchDetectorState(statusMessage = "Could not start microphone")
            return
        }

        val recorder = AudioRecord(
            MediaRecorder.AudioSource.DEFAULT,
            sampleRate,
            AudioFormat.CHANNEL_IN_MONO,
            AudioFormat.ENCODING_PCM_FLOAT,
            max(minimumBuffer, frameSize * Float.SIZE_BYTES * 2),
        )
        if (recorder.state != AudioRecord.STATE_INITIALIZED) {
            recorder.release()
            _state.value = PitchDetectorState(statusMessage = "Could not start microphone")
            return
        }

        audioRecord = recorder
        smoothedFrequency = null
        missedFrameCount = 0
        running = true
        recorder.startRecording()
        _state.value = PitchDetectorState(isRunning = true, statusMessage = "Listening...")

        thread(name = "WindBrassPitchDetector", isDaemon = true) {
            val samples = FloatArray(frameSize)
            while (running && generation.get() == session) {
                val count = recorder.read(samples, 0, samples.size, AudioRecord.READ_BLOCKING)
                if (count > 32 && generation.get() == session) {
                    publishStableFrequency(detectPitch(samples, count, sampleRate.toDouble(), targetFrequency))
                }
            }
        }
    }

    fun stop() {
        generation.incrementAndGet()
        running = false
        audioRecord?.let { recorder ->
            runCatching { recorder.stop() }
            recorder.release()
        }
        audioRecord = null
        smoothedFrequency = null
        missedFrameCount = 0
        _state.value = PitchDetectorState()
    }

    private fun publishStableFrequency(frequency: Double?) {
        if (frequency == null || frequency <= 0) {
            missedFrameCount++
            val held = smoothedFrequency.takeIf { missedFrameCount <= 5 }
            _state.value = PitchDetectorState(
                detectedFrequency = held,
                isRunning = running,
                statusMessage = if (held != null) "Holding tone" else "Listening...",
            )
            return
        }

        missedFrameCount = 0
        smoothedFrequency = smoothedFrequency?.let { previous ->
            val distanceInCents = abs(1200 * log2(frequency / previous))
            if (distanceInCents < 700) previous * 0.72 + frequency * 0.28 else frequency
        } ?: frequency
        _state.value = PitchDetectorState(smoothedFrequency, running, "Pitch detected")
    }

    companion object {
        internal fun detectPitch(samples: FloatArray, count: Int, sampleRate: Double, targetFrequency: Double): Double? {
            if (count <= 32 || sampleRate <= 0 || targetFrequency <= 0) return null
            var energy = 0.0
            var sum = 0.0
            for (index in 0 until count) {
                val value = samples[index].toDouble()
                energy += value * value
                sum += value
            }
            if (sqrt(energy / count) <= 0.01) return null

            val mean = sum / count
            val centered = DoubleArray(count) { samples[it] - mean }
            val minLag = max(8, (sampleRate / 1_200.0).toInt())
            val maxLag = min(count - 2, (sampleRate / 70.0).toInt())
            if (minLag >= maxLag) return null

            val scores = DoubleArray(maxLag + 1)
            var bestLag = minLag
            var bestScore = Double.NEGATIVE_INFINITY
            for (lag in minLag..maxLag) {
                var correlation = 0.0
                var energyA = 0.0
                var energyB = 0.0
                for (index in 0 until count - lag) {
                    val a = centered[index]
                    val b = centered[index + lag]
                    correlation += a * b
                    energyA += a * a
                    energyB += b * b
                }
                val denominator = sqrt(energyA * energyB)
                if (denominator <= 0) continue
                val score = correlation / denominator
                scores[lag] = score
                if (score > bestScore) {
                    bestScore = score
                    bestLag = lag
                }
            }
            if (bestScore <= 0.45) return null

            var refinedLag = bestLag.toDouble()
            if (bestLag > minLag && bestLag < maxLag) {
                val curvature = scores[bestLag - 1] - 2 * scores[bestLag] + scores[bestLag + 1]
                if (abs(curvature) > 0.000_001) {
                    refinedLag += (0.5 * (scores[bestLag - 1] - scores[bestLag + 1]) / curvature).coerceIn(-0.5, 0.5)
                }
            }
            return sampleRate / refinedLag
        }
    }
}
